pkg/apis/pipeline/v1alpha1/taskrun_validation_test.go

```diff
@@ -0,0 +1,313 @@
+			if d := cmp.Diff(err.Error(), ts.want.Error()); d != "" {
+				t.Errorf("Validate/%s (-want, +got) = %v", ts.name, d)
+			}
+		})
+	}
+}
+
+func TestTaskRunSpecValidate(t *testing.T) {
+	tests := []struct {
+		name    string
+		spec    *TaskRunSpec
+		wantErr *apis.FieldError
+	}{
+		{
+			name:    "invalid taskspec",
+			spec:    &TaskRunSpec{},
+			wantErr: apis.ErrMissingField("spec"),
+		},
+		{
+			name: "invalid taskref",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{},
+				Trigger: TaskTrigger{
+					TriggerRef: TaskTriggerRef{
+						Type: TaskTriggerTypePipelineRun,
+					},
+				},
+			},
+			wantErr: apis.ErrMissingField("spec.taskref.name"),
+		},
+		{
+			name: "invalid taskref",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Trigger: TaskTrigger{
+					TriggerRef: TaskTriggerRef{
+						Type: "wrongtype",
+					},
+				},
+			},
+			wantErr: apis.ErrInvalidValue("wrongtype", "spec.trigger.triggerref.type"),
+		},
+		{
+			name: "valid task trigger run type",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Trigger: TaskTrigger{
+					TriggerRef: TaskTriggerRef{
+						Type: TaskTriggerTypePipelineRun,
+					},
+				},
+			},
+		},
+		{
+			name: "valid task inputs",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Inputs: TaskRunInputs{
+					Params: []Param{{
+						Name:  "name",
+						Value: "value",
+					}},
+					Resources: []PipelineResourceVersion{{
+						Version: "testv1",
+						ResourceRef: PipelineResourceRef{
+							Name: "testresource",
+						},
+					}},
+				},
+			},
+		},
+		{
+			name: "invalid task inputs",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Inputs: TaskRunInputs{
+					Resources: []PipelineResourceVersion{{
+						Version: "testv1",
+						ResourceRef: PipelineResourceRef{
+							Name: "testresource",
+						},
+					}, {
+						Version: "testv1",
+						ResourceRef: PipelineResourceRef{
+							Name: "testresource",
+						},
+					}},
+				},
+			},
+			wantErr: apis.ErrMultipleOneOf("spec.Inputs.Resources.Name"),
+		},
+		{
+			name: "invalid task input params",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Inputs: TaskRunInputs{
+					Resources: []PipelineResourceVersion{{
+						Version: "testv1",
+						ResourceRef: PipelineResourceRef{
+							Name: "testresource",
+						},
+					}},
+					Params: []Param{{
+						Name:  "name",
+						Value: "value",
+					}, {
+						Name:  "name",
+						Value: "value",
+					}},
+				},
+			},
+			wantErr: apis.ErrMultipleOneOf("spec.inputs.params"),
+		},
+		{
+			name: "invalid task output type",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Outputs: Outputs{
+					Resources: []TaskResource{{
+						Type: "invalidtype",
+					}},
+				},
+			},
+			// TODO(shashwathi): wrong error msg
```

Comment:
can we fix the error msg?

Reply:
Damn I was suppose to fix it before sending PR :D